src/stores/modelStore.ts

```diff
@@ -0,0 +1,141 @@
+    }
+  }
+  return null
+}
+
+/** Defines and holds metadata for a model */
+export class ComfyModelDef {
+  /** Proper filename of the model */
+  name = ref('')
+  /** Directory containing the model, eg 'checkpoints' */
+  directory = ref('')
+  /** Title / display name of the model, sometimes same as the name but not always */
+  title = ref('')
+  /** Metadata: architecture ID for the model, such as 'stable-diffusion-xl-v1-base' */
+  architecture_id = ref('')
+  /** Metadata: author of the model */
+  author = ref('')
+  /** Metadata: resolution of the model, eg '1024x1024' */
+  resolution = ref('')
+  /** Metadata: description of the model */
+  description = ref('')
+  /** Metadata: usage hint for the model */
+  usage_hint = ref('')
+  /** Metadata: trigger phrase for the model */
+  trigger_phrase = ref('')
+  /** Metadata: tags list for the model */
+  tags: Ref<string[]> = ref([])
+  /** Metadata: image for the model */
+  image = ref('')
+  /** Whether the model metadata has been loaded from the server, used for `load()` */
+  has_loaded_metadata = ref(false)
+
+  constructor(name: string, directory: string) {
+    this.name.value = name
+    this.title.value = name
+    this.directory.value = directory
+  }
+
+  /** Loads the model metadata from the server, filling in this object if data is available */
+  async load(): Promise<void> {
+    if (this.has_loaded_metadata.value) {
+      return
+    }
+    const metadata = await api.viewMetadata(
+      this.directory.value,
+      this.name.value
+    )
+    if (!metadata) {
+      return
+    }
+    this.title.value =
+      _findInMetadata(
+        metadata,
+        'modelspec.title',
+        'title',
+        'display_name',
+        'name'
+      ) || this.name.value
+    this.architecture_id.value =
+      _findInMetadata(metadata, 'modelspec.architecture', 'architecture') || ''
+    this.author.value =
+      _findInMetadata(metadata, 'modelspec.author', 'author') || ''
+    this.description.value =
+      _findInMetadata(metadata, 'modelspec.description', 'description') || ''
+    this.resolution.value =
+      _findInMetadata(metadata, 'modelspec.resolution', 'resolution') || ''
+    this.usage_hint.value =
+      _findInMetadata(metadata, 'modelspec.usage_hint', 'usage_hint') || ''
+    this.trigger_phrase.value =
+      _findInMetadata(metadata, 'modelspec.trigger_phrase', 'trigger_phrase') ||
+      ''
+    this.image.value =
+      _findInMetadata(
+        metadata,
+        'modelspec.thumbnail',
+        'thumbnail',
+        'image',
+        'icon'
+      ) || ''
+    const tagsCommaSeparated =
+      _findInMetadata(metadata, 'modelspec.tags', 'tags') || ''
+    this.tags.value = tagsCommaSeparated.split(',').map((tag) => tag.trim())
+    this.has_loaded_metadata.value = true
+  }
+}
+
+/** Model store for a folder */
+export class ModelStore {
+  models: Record<string, ComfyModelDef> = reactive({})
+
+  constructor(directory: string, models: string[]) {
```

Comment:
1. Instead of using a class for ModelStore, it would be better to incorporate this logic directly into the Pinia store. This would make the store more self-contained and easier to manage.
2. The reactive wrapper is not necessary when defining state in a Pinia store, as Pinia automatically makes the state reactive.

Here's how you could refactor this to better align with Pinia best practices:
```ts
import { defineStore } from 'pinia'
import { ComfyModelDef } from './ComfyModelDef' // Assume this is moved to a separate file

export const useModelStore = defineStore('modelStore', {
  state: () => ({
    modelStoreMap: {} as Record<string, Record<string, ComfyModelDef>>
  }),
  actions: {
    async getModelsInFolderCached(folder: string) {
      if (folder in this.modelStoreMap) {
        return this.modelStoreMap[folder]
      }
      // TODO: needs a lock to avoid overlapping calls
      const models = await api.getModels(folder)
      if (!models) {
        return null
      }
      this.modelStoreMap[folder] = {}
      for (const model of models) {
        this.modelStoreMap[folder][model] = new ComfyModelDef(model, folder)
      }
      return this.modelStoreMap[folder]
    },
    async loadModelMetadata(folder: string, modelName: string) {
      if (this.modelStoreMap[folder]?.[modelName]) {
        await this.modelStoreMap[folder][modelName].load()
      }
    },
    clearCache() {
      this.modelStoreMap = {}
    }
  }
})
```

Reply:
Pinia best practices increasingly seem to be directly contradictory to general programming best practices. "Don't use a subclass because pinia can't do that, just shove it all into one" seems more like an argument against pinia than anything else

Reply:
I would still recommend you follow the direct storage type `Record<string, Record<string, ComfyModelDef>>`. The store is essentially storing a collection of the ModelDef, and the aux `ModelStore` is just provide a way to access ModelDef by name. You can define getter (computed state) later to add other ways to access data `Record<modelName, Record<modelDir, ComfyModelDef>>`. Following your approach, we would need a ModelStoreByDir to manage that ownership.

The first principle is that Pinia store should own all objects it manages and the method to access them.